src/utils.rs

```diff
@@ -241,38 +241,80 @@ pub fn get_encryption_key(enc_params: &EncryptionParams) -> Result<&[u8]> {
+/// The only problem here we need to fix is that the SecurityBufferType::Stream only has the decrypted data, not the token.
+/// This is a temporary fix until we implement the correct behavior.
+pub fn save_decrypted_data<'a>(decrypted: &'a [u8], buffers: &'a mut [DecryptBuffer]) -> Result<()> {
+    let buffer = DecryptBuffer::find_buffer_mut(buffers, SecurityBufferType::Stream);
+
+    let mut stream_buffer_inner = None;
+    // If there is a stream buffer, then we will write the decrypted data into it.
+    if let Ok(stream_buffer) = buffer {
+        if stream_buffer.data().len() < decrypted.len() {
             return Err(Error::new(
-                ErrorKind::DecryptFailure,
-                "Decrypted data can not be larger then encrypted one.",
+                ErrorKind::EncryptFailure,
+                format!(
+                    "Decrypted data length ({}) does not match the stream buffer length ({})",
+                    decrypted.len(),
+                    stream_buffer.data().len()
+                ),
             ));
         }
-        buffer
+
+
+        let mut inner_stream_buffer = stream_buffer.take_data();
+
+        stream_buffer_inner = Some(DataBuffer {
+            data: inner_stream_buffer.as_mut_ptr(),
+            size: decrypted.len(),
+        });
+
+        inner_stream_buffer = &mut inner_stream_buffer[..decrypted.len()];
+        inner_stream_buffer.copy_from_slice(&decrypted[..]);
+        stream_buffer.set_data(inner_stream_buffer)?;
+
     };
 
-    let mut data_buffer = buffer.take_data();
+    let buffer = DecryptBuffer::find_buffer_mut(buffers, SecurityBufferType::Data);
+    // If there is a data buffer.
+    if let Ok(data_buffer) = buffer {
+        // and if the data buffer is large enough, then we will write the decrypted data into it as well.
+        if data_buffer.data().len() > decrypted.len() {
+            let mut inner_data_buffer = data_buffer.take_data();
+            inner_data_buffer = &mut inner_data_buffer[..decrypted.len()];
+            inner_data_buffer.copy_from_slice(&decrypted[..]);
+            data_buffer.set_data(inner_data_buffer)?;
+        // Otherwise, if the stream buffer is present, then we set the data buffer to the inner pointer of the stream buffer.
+        // This is inherently unsafe, but it is the behavior of the Windows SSPI.
+        }else if let Some(stream_buffer) = stream_buffer_inner {
+            let DataBuffer {
+                data: stream_inner_ptr,
+                size: stream_inner_size,
+            } = stream_buffer;
 
-    // Safe: checked above.
-    data_buffer = &mut data_buffer[0..decrypted.len()];
-    data_buffer.copy_from_slice(decrypted);
+            unsafe {
+                let data_slice = std::slice::from_raw_parts_mut(stream_inner_ptr, stream_inner_size);
+                data_buffer.set_data(data_slice)?;
+            }
```

Comment:
suggestion: Keep the unsafe block as small as possible.

```rust
            let data_slice = unsafe {
                // SAFETY: This is safe because…
                std::slice::from_raw_parts_mut(stream_inner_ptr, stream_inner_size)
            };
            data_buffer.set_data(data_slice)?;
```

Reply:
updated